### ISSUE TITLE
selecting closest point in 1-D

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#21693]
Given a range of values, a common approach to finding the closest value to some reference is to calculate the absolute values of the differences and then select the one with the minimum difference. We can do better inthe case of handling rationals if we avoid all the fraction mechanics of calculating the differences and reducing the results. 

```python
def nd(f):
    return f.numerator, f.denominator

def closer(p,q,a,b,c,d):
    # if |a/b - p/q| < |c/d - p/q| then
    # (a/b-p/q)**2 < (c/d-p/q)**2
    # (a/b-p/q)**2 - (c/d-p/q)**2 < 0
    # >>> n,d=((a/b-p/q)**2 - (c/d-p/q)**2).as_numer_denom()
    # >>> n.expand()
    # a**2*d**2*q**4 - 2*a*b*d**2*p*q**3 - b**2*c**2*q**4 + 2*b**2*c*d*p*q**3
    # >>> d.expand()
    # b**2*d**2*q**4
    # >>> factor(n)
    # -q**3*(a*d - b*c)*(-a*d*q - b*c*q + 2*b*d*p)
    # denom and q are positive so ignore
    ad = a*d
    bc = b*c
    cmp = lambda a,b: 1 if a>b else (0 if a==b else -1)
    disc = cmp(bc, ad)*cmp(2*b*d*p, q*(bc + ad))
    if disc < 0:
        return a,b
    return c, d

def nearest(t, f):
    p,q=nd(t)
    f = [nd(i) for i in f]
    a,b=f[0]
    for i in range(len(f)):
        a,b = closer(p,q,a,b,*f[i])
    return F(a,b)

from fractions import Fraction as F
from time import time
f = [F(random()).limit_denominator(10**4) for i in range(10**4)]
t=f.pop()
d=time(); ans=min(list(zip([abs(i-t) for i in f], f)))[1]; ok=time()-d
print(ans,ok)
d=time(); ans=nearest(t,f); ok=time()-d
print(ans, ok)
```
The results indicate that a significant savings is possible (somewhere in the 3-10X range):
```
1891/3656 0.276975154876709
1891/3656 0.027315139770507812
```
Perhaps a `Rational.nearest(*vals)` would be worth implementing. This also might work as a refinement of results narrowed down by using floats instead of rational (i.e. converting each value to a float) before calculating the absolute value; doing that is faster than using `nearest` but is limited in precision, so ties could be broken with the `nearest` function.   But if there are only a few ties them perhaps even this is an insignificant improvement.

**Comment by oscarbenjamin** [comment]
This implementation seems a little faster:
```python
import bisect

def nearest2(t, f):
    f = sorted(f, key=lambda r: (float(r), r))
    i = bisect.bisect(f, t)
    if i and abs(f[i-1] - t) < abs(f[i] - t):
        i -= 1
    return f[i]
```

**Comment by oscarbenjamin** [comment]
I'm not sure what this would be useful for though...

**Comment by smichr** [comment]
> I'm not sure what this would be useful for though...

in collision detection (of segments in a 2-D plane) one has to find values that are within some range. I am not sure what sort of optimizations, within things like the sweep-line algorithm, are done on the computational side. When I tried to implement this it was very slow in SymPy and I just gave up and left those sorts of calculations to apps specifically made for that purpose.
